coreax/coreset.py

```diff
@@ -122,12 +124,43 @@ class Coresubset(Coreset[_Data], Generic[_Data]):
     @property
     def coreset(self) -> Data:
         """Materialise the coresubset from the indices and original data."""
-        coreset_nodes = self.pre_coreset_data.data[self.unweighted_indices]
-        return Data(coreset_nodes, self.nodes.weights)
+        if isinstance(self.pre_coreset_data, SupervisedData):
+            coreset_data = self.pre_coreset_data.data[self.unweighted_indices]
+            coreset_supervision = self.pre_coreset_data.supervision[
+                self.unweighted_indices
+            ]
+            return SupervisedData(
+                data=coreset_data,
+                supervision=coreset_supervision,
+                weights=self.nodes.weights,
+            )
+        coreset_data = self.pre_coreset_data.data[self.unweighted_indices]
+        return Data(data=coreset_data, weights=self.nodes.weights)
 
     @property
     def unweighted_indices(self) -> Shaped[Array, " n"]:
         """Unweighted Coresubset indices - attribute access helper."""
         return jnp.squeeze(self.nodes.data)
 
+    def reverse(self) -> "Coresubset":
+        """
+        Return the coresubset with the order of its coreset indices reversed.
+
+        Convenience method for changing the order of coreset index refinement.
+        """
+        return Coresubset(
+            Data(jnp.flip(self.unweighted_indices)), self.pre_coreset_data
+        )
+
+    def permute(self, random_key: KeyArrayLike) -> "Coresubset":
+        """
+        Return the coresubset with the order of its coreset indices permuted.
+
+        Convenience method for changing the order of coreset index refinement.
+        """
+        return Coresubset(
+            Data(jr.permutation(random_key, self.unweighted_indices)),
+            self.pre_coreset_data,
+        )
+
```

Comment:
These methods would replace the old functionality of `RefineReverse` and add additional functionality to randomise order of refinement.  Because the methods in  `RefinementSolver.refine(...)` work in the order that the coreset indices are inputted you can get stuck in local minima where additional recursive calls to refinement are stuck. By reversing or permuting the order of the coreset indices you can jump out of these local minima at no cost.

Thanks for the general spots!

Reply:
I'm getting pylint complaints with Self but not with "Coresubset"

Reply:
![image](https://github.com/gchq/coreax/assets/167978891/a73bdfc1-28e3-4be0-8b80-466274cbbda2)

Reply:
I think the type hinting issue can be fixed by using `eqx.tree_at`. This way the generic types will be preserved and the type hint complain should go away. In general, I would prefer the user to be responsible for performing these index re-orderings as It only requires making a copy of the coreset with updated nodes. For example 
```python
reversed_coreset = eqx.tree_at(lambda x: x.nodes, coreset, coreset.nodes[::-1])
permuted_indices = jr.choice(key, len(coreset))
permuted_coreset = eqx.tree_at(lambda x: x.nodes, coreset, coreset.nodes[permuted_indices])
```
This allows the user to perform arbitrary re-indexing of the `nodes`, which is more powerful than what is exposed by the convenience interfaces, and avoids any ambiguity about what exactly is being permuted or reversed.

I'll add that from a design standpoint, the responsibility of handling the problem of local-minima should be that of the refinement solver itself, or better yet, some "controller" that wraps/interfaces with the solver. Unless there is an obvious demand from users to have these convenience interfaces, I would argue that the more appropriate solution here would be to provide a means to properly control the refinement solvers to avoid local-minima (or at least to enumerate a set of local-minima). This would certainly require a separate ticket though.

Reply:
Happy to remove these from the PR.